systemd create timer option

harishanand95 opened the pull request:
Work left to complete :
- [x] Invalid input message shown for inputs.
- [x] Repeat Hourly, Daily, Weekly, Monthly, Yearly and dont repeat option works.
- [x] Tests cases

Fixes left:
- [x] Create Timer button shown in all services tab than just timer (will be fixed when #4503 merges).
- [x] Repeat Yearly's datepicker is positioned incorrectly.
- [x] “Invalid input” message causes change of position of boot timer’s elements.

andreasn commented:
Good start!
There is a couple of styling and spacing errors, but those should be quick and easy to fix.

![screenshot from 2016-07-01 16-39-42](https://cloud.githubusercontent.com/assets/209895/16524955/72b935ae-3faa-11e6-9da7-3589098b9157.png)

Compare with
![screenshot from 2016-07-01 16-39-12](https://cloud.githubusercontent.com/assets/209895/16525000/972a071a-3faa-11e6-9be3-9b1352974a48.png)

* The spacing between the rows is too large.
* The labels are black instead of gray.
* The spacing between the labels on the left and the inputs on the right is too big.

andreasn commented:
Should "Service file name" be "Service name" perhaps?

andreasn commented:
Mockup for comparison.
![Mockup](https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/services/timers.png)

andreasn commented:
Instead of putting "At" on it's own row, I think it would work better to put it on the same line as the dropdown.

harishanand95 commented:
Updated design changes (I think now it more follows the mock-ups) :
![screenshot from 2016-07-04 16-25-37](https://cloud.githubusercontent.com/assets/7456909/16558673/cf0f3d56-4204-11e6-8c46-9d2acb8260c3.png)
![screenshot from 2016-07-04 16-25-43](https://cloud.githubusercontent.com/assets/7456909/16558674/d06ddb62-4204-11e6-8da5-dc6312e55949.png)
![screenshot from 2016-07-04 16-25-53](https://cloud.githubusercontent.com/assets/7456909/16558676/d25d70cc-4204-11e6-81f5-983821c7b59f.png)
![screenshot from 2016-07-04 16-25-58](https://cloud.githubusercontent.com/assets/7456909/16558683/d39c3522-4204-11e6-9eea-c469663a6600.png)

andreasn commented:
Good improvements. This is starting to shape up really well!

A couple of minor issues:
 * "Run after system boot at X seconds" doesn't sound quite right. I think "Run _after_ system boot after X seconds" would work better.
 * It would look a bit nicer if the "at/after" label, the box and the seconds dropdown would be right aligned.
* If the "at/after" box is left empty, does that mean the delay is zero? It would be good to fill the input with that number in that case.
* Could the + for adding another case be on the same row as the x that removes it? That would make it consistent with how it works in the Container Run dialog.

harishanand95 commented:
@andreasn fixed all the issues you have mentioned.
other fixes in the commit are : 
* Create Timer button is shown in timer tab only.
*  “Invalid input” message doesnt change of position of boot_time elements.

fixes still left : 
* Repeat Yearly's datepicker menu is positioned incorrectly.

harishanand95 commented:
@andreasn I have corrected the position of datepicker in repeat yearly option.  
![screenshot from 2016-07-06 17-16-31](https://cloud.githubusercontent.com/assets/7456909/16616835/7ef766b4-439d-11e6-8d4c-554cec1b3493.png)

dperpeet commented:
> "Run after system boot after X seconds" would work better.

How about "run X seconds after system boot"?

andreasn commented:
> How about "run X seconds after system boot"?
Even though it forms a better sentence, it would make the dropdown jump around.

andreasn commented:
If I first create one timer, then decides to create another, the values from the old timer is still present in the timer creation dialog. It needs to be cleared out on exit.

andreasn commented:
The error message "within 0-59" sounds a bit odd. Maybe "Needs to be between 0-59"
Or is that too long of a string to fit?

andreasn commented:
Maybe this is a limitation of the calendar widget, but it feels a bit strange to be able to run a timer in the past. Not a biggie though.

andreasn commented:
![screenshot from 2016-07-07 19-11-49](https://cloud.githubusercontent.com/assets/209895/16662228/c0904fb2-4476-11e6-94d7-acdcbe6253e2.png)

andreasn commented:
Maybe the time errors could be displayed like this instead:
![screenshot from 2016-07-07 19-11-49](https://cloud.githubusercontent.com/assets/209895/16662605/7cdb4ec8-4478-11e6-9585-cc78f773cfa4.png)

harishanand95 commented:
@andreasn I have updated the error messages. Bootstrap datepicker is difficult to work with! I needed a hack to get its position correct. :(
Here are the screen shots:
![screenshot from 2016-07-11 11-36-09](https://cloud.githubusercontent.com/assets/7456909/16721331/0d829900-475c-11e6-9376-f8db1336d57e.png)
![screenshot from 2016-07-11 11-36-18](https://cloud.githubusercontent.com/assets/7456909/16721332/0f1bae6e-475c-11e6-8087-6289cc34c515.png)

andreasn commented:
It seems the alignment for the labels for "Service name", "Description" and "Command" gets off once they get a error message. They get pushed too far down.

![screenshot from 2016-07-11 17-08-40](https://cloud.githubusercontent.com/assets/209895/16735636/236b5a58-478a-11e6-8dcf-e645545ab084.png)

Filling an entry has the opposite effect. The labels end up too far up.
![screenshot from 2016-07-11 17-11-31](https://cloud.githubusercontent.com/assets/209895/16735754/8eaf20c4-478a-11e6-97a5-555f5d8ac022.png)

andreasn commented:
I wonder if it makes sense to spread the Repeat dropdown all across to line it up with the elements below better. Should be just a matter of removing the div#drop_repeat css

dperpeet commented on pkg/systemd/init.js:
```diff
@@ -267,7 +268,10 @@ define([
                 Current_State: _("State")
             };
-
+            if (~pattern.indexOf("timer"))
```
I find the bit-operator `~` misleading here.

dperpeet commented on pkg/systemd/init.js:
```diff
@@ -823,4 +827,490 @@ define([
+        else
+            $(this).parents("tr").next().hide();
+    });
+    /* HACK - bootstrap datepicker positions itself incorrectly on modals
```
Is there an upstream issue for this?
Usually we try to ensure that a HACK is only temporary: either a fix is already implemented and will be available in a future version or at least the issue is known upstream and will hopefully be fixed some day.

dperpeet commented on pkg/systemd/init.js:
```diff
@@ -823,4 +827,490 @@ define([
+            $("#hr").removeClass("has-error");
+            $("#min").removeClass("has-error");
+        }
+        switch(val) {
```
Instead of a switch we could check for `0` and map the rest like { 60: 'Repeat Hourly' } et cetera.

dperpeet commented on pkg/systemd/init.js:
```diff
@@ -823,4 +827,490 @@ define([
+                $("#close_button").hide();
+                $("#hr").val("00");
+                $("#min").val("00");
+                timer_unit.repeat = "Don't Repeat";
```
Will these phrases be translated later? If they aren't, they should be translated here: `_("Don't Repeat")`

dperpeet commented on pkg/systemd/init.js:
```diff
@@ -823,4 +827,490 @@ define([
+            case 525600 : timer_unit.repeat = "Repeat Yearly";// 60*24*365min
+                break;
+        }
+        if (val) {
```
this can be combined with the above into and `if-else` on `val`

dperpeet commented on pkg/systemd/init.js:
```diff
@@ -823,4 +827,490 @@ define([
+            date:moment().format("YYYY-MM-DD")
+        };
+
+        if (timer_unit.repeat == "Repeat Hourly") {
```
Many of the blocks in this `if-else` construct look similar. Can they be combined? You could use `||` in the conditions.

dperpeet commented on pkg/systemd/init.js:
```diff
@@ -823,4 +827,490 @@ define([
+    });
+
+    function display_repeat() {
+        if (timer_unit.repeat == "Repeat Hourly") {
```
These blocks look very similar again, mostly. Can this be simplified?

dperpeet commented on pkg/systemd/services.html:
```diff
@@ -136,6 +138,134 @@
+          <span class="caret"></span>
+        </button>
+        <ul class="dropdown-menu vertical-scroll">
+          <li value="1"><a>1st</a></li>
```
@andreasn how would we translate this?

andreasn replied:
Good question. Yes, I think translatable=yes could work.
So how would it look in German? In Swedish it would be:
1:a, 2:a, 3:e, 4:e, 5:e etc. and maybe that's just asking for trouble.
1, 2, 3, 4, 5 etc. would work for me too.

dperpeet replied:
> Good question. Yes, I think translatable=yes could work.
> So how would it look in German? In Swedish it would be:
> 1:a, 2:a, 3:e, 4:e, 5:e etc. and maybe that's just asking for trouble.
> 1, 2, 3, 4, 5 etc. would work for me too.

I think I would prefer to simply use 1, 2, 3, ... instead of increasing translation difficulty

dperpeet commented on pkg/systemd/services.html:
```diff
@@ -136,6 +138,134 @@
+          <span class="caret"></span>
+        </button>
+        <ul class="dropdown-menu">
+          <li value="1"><a>Monday</a></li>
```
@andreasn how do we translate this? translatable?

dperpeet commented on pkg/systemd/services.html:
```diff
@@ -196,5 +326,147 @@ <h4 class="modal-title" translatable="yes">Error</h4>
+    <div class="modal-dialog cockpit-timer-modal-md">
+      <div class="modal-content">
+        <div class="modal-header">
+          <h4 class="modal-title">Create Timers</h4>
```
All tags with English text should be marked as translatable.

dperpeet commented:
This pull request is making really good progress!

harishanand95 commented:
@dperpeet @andreasn   I have corrected the issues pointed out specifically, added translatable=yes to all html english words and simplified code of repeat options.
* I haven't added translatable=yes for contents that lies within mustache. This includes texts like 1st, 2nd, 3rd, Monday, Tuesday et cetera. Should these be passed from within init.js in _() or translatable=yes?
* @andreasn could u show me where to add the popover-information circle icon? Is it to be placed near 29th-31st dropdown only or shown for all cases somewhere else? 
![screenshot from 2016-07-25 00-57-50](https://cloud.githubusercontent.com/assets/7456909/17085950/e2e09ace-5202-11e6-8cf6-b637498df3b8.png)

andreasn commented:
> could u show me where to add the popover-information circle icon?

I could be remembering wrong, but I thought we said to use a warning for when you select 31st only.

harishanand95 commented:
@andreasn @dperpeet Is "better avoid end of month days like 31st" okay? like you have said @andreasn I gave color: #4d5258; which was the color of the text "Repeat Monthly".

![screenshot from 2016-08-01 16-53-33](https://cloud.githubusercontent.com/assets/7456909/17294284/7d5f5e12-5813-11e6-8239-0d8ace6b11f2.png)

@dperpeet I have added the tests: 
The tests added now  has test cases for repeat hourly, daily, weekly, monthly and yearly and error inputs.
The test doesn't check for a boot timer and a timer that doesn't repeat. I will add these soon.

andreasn commented:
> better avoid end of month days like 31st

This doesn't sound quite right to me. It's not a full sentence.
What about:
"this date only occur every other month"

andreasn commented:
I'm also thinking we could actually leave out the 31st all together, and possibly add an entry for Feb 28th whenever you schedule something on the 30th.

larsu commented:
> This doesn't sound quite right to me. It's not a full sentence.

I agree. I think we should also include _why_ it won't occur ever month. How about (as a warning, in red): "This day doesn't exist in all months. The timer will only be executed in months that have a 31st."

Of course, this is only if systemd really behaves that way. I haven't tested it myself.

I think a better design would be to simply move the event to the last day of the month for each month and show a non-warning label that this is done. That needs support in systemd. Could you please open a bug there?

harishanand95 commented:
@larsu @andreasn when I mentioned this in #systemd they told such days will be skipped and asked me to avoid end of the month days. I think we should do a similar warning like you said "This day doesn't exist in all months. The timer will only be executed in months that have a 31st." Is red color okay? @andreasn 
I will file an issue about this in systemd

harishanand95 commented:
![screenshot from 2016-08-05 14-10-09](https://cloud.githubusercontent.com/assets/7456909/17432130/3283c38c-5b1b-11e6-8c94-b67d53260fe5.png)
@andreasn this is the info shown for non-superuser accounts and for 31st. okay?
![screenshot from 2016-08-05 14-46-18](https://cloud.githubusercontent.com/assets/7456909/17432175/7771f400-5b1b-11e6-8d21-aaaf0140b681.png)

andreasn commented:
Both looks good to me. Thanks for fixing that!
Might want to do a new line for "The timer will only...". I think that will make it flow better.

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +188,268 @@ WantedBy=default.target
+        m.spawn("sync && sync && sync && sleep 0.1 && reboot", "reboot")
+        m.wait_reboot()
+        m.start_cockpit()
+        self.assertIn("hello", m.execute("cat /tmp/hello"))
```
The timer runs 10 seconds after boot, but here you assert that the text is present after cockpit started. To avoid a race condition, you should probably safely wait for the text to appear.
Maybe something like this would work:
`m.execute("while [ ! -f /tmp/hello ] ; do sleep 0.5; done")`

dperpeet replied:
and once that exists, you can make sure the text is in there - basically add the wait before this line

dperpeet replied:
this execute command would need to be enclosed in a timer, e.g.:
`with Timeout(seconds=15, error_message="Timeout while waiting for boot timer to run"):`

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        self.assertIn("Wed 2020-01-01 16:00", m.execute("LC_ALL=C systemctl list-timers"))
+        m.execute("timedatectl set-time '2020-01-01 16:10:00'")
+        b.wait_present(svc_sel('yearly_timer.timer'))
+        self.assertIn("Fri 2021-01-01 01:22", m.execute("LC_ALL=C systemctl list-timers"))
```
This looks like a race condition ([failed test](http://fedorapeople.org/groups/cockpit/logs/pull-4645-a1023602-verify-centos-7/log.html)) - I think systemd needs some time to update this file. Why not wait for this info to be properly displayed in the cockpit ui? Then you can use the standard wait functionality for the text to be updated.

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        self.assertIn("Mon 2020-01-06 14:12", m.execute("LC_ALL=C systemctl list-timers"))
+        m.execute("timedatectl set-time '2020-01-07 00:00:00'")
+        b.wait_present(svc_sel('monthly_timer.timer'))
+        self.assertIn("Wed 2020-01-08 21:12", m.execute("LC_ALL=C systemctl list-timers"))
```
see above, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        # checks if timer runs on next month February 2020 on same dates
+        m.execute("timedatectl set-time '2020-02-05 00:00:00'")
+        b.wait_present(svc_sel('monthly_timer.timer'))
+        self.assertIn("Thu 2020-02-06 14:12", m.execute("LC_ALL=C systemctl list-timers"))
```
see above, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        # checks if timer runs on 8th March 2020 at 21:12
+        m.execute("timedatectl set-time '2020-03-07 00:00:00'")
+        b.wait_present(svc_sel('monthly_timer.timer'))
+        self.assertIn("Sun 2020-03-08 21:12", m.execute("LC_ALL=C systemctl list-timers"))
```
see above, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        m.execute("systemctl start weekly_timer.timer")
+        b.wait_present(svc_sel('weekly_timer.timer'))
+        b.wait_in_text(svc_sel('weekly_timer.timer'), "Weekly timer")
+        self.assertIn("Sun 2020-03-08 20:12", m.execute("LC_ALL=C systemctl list-timers"))
```
see above, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        self.assertIn("Sun 2020-03-08 20:12", m.execute("LC_ALL=C systemctl list-timers"))
+        m.execute("timedatectl set-time '2020-03-09 00:00:00'")
+        b.wait_present(svc_sel('weekly_timer.timer'))
+        self.assertIn("Fri 2020-03-13 12:45", m.execute("LC_ALL=C systemctl list-timers"))
```
see above, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        # checks if timer runs on next week's Friday and Sunday
+        m.execute("timedatectl set-time '2020-03-14 00:00:00'")
+        b.wait_present(svc_sel('weekly_timer.timer'))
+        self.assertIn("Sun 2020-03-15 20:12", m.execute("LC_ALL=C systemctl list-timers"))
```
see above, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        self.assertIn("Sun 2020-03-15 20:12", m.execute("LC_ALL=C systemctl list-timers"))
+        m.execute("timedatectl set-time '2020-03-17 00:00:00'")
+        b.wait_present(svc_sel('weekly_timer.timer'))
+        self.assertIn("Fri 2020-03-20 12:45", m.execute("LC_ALL=C systemctl list-timers"))
```
see above, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        m.execute("systemctl start daily_timer.timer")
+        b.wait_present(svc_sel('daily_timer.timer'))
+        b.wait_in_text(svc_sel('daily_timer.timer'), "Daily timer")
+        self.assertIn("Tue 2020-03-17 02:40", m.execute("LC_ALL=C systemctl list-timers"))
```
see above, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

or is this one alleviated by starting it on the command line?

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        self.assertIn("Tue 2020-03-17 02:40", m.execute("LC_ALL=C systemctl list-timers"))
+        m.execute("timedatectl set-time '2020-03-17 03:00:00'")
+        b.wait_present(svc_sel('daily_timer.timer'))
+        self.assertIn("Tue 2020-03-17 21:15", m.execute("LC_ALL=C systemctl list-timers"))
```
see above, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        # checks if timer runs on friday 2020-04-10 at 02:40 and 21:15
+        m.execute("timedatectl set-time '2020-04-10 00:00:00'")
+        b.wait_present(svc_sel('daily_timer.timer'))
+        self.assertIn("Fri 2020-04-10 02:40", m.execute("LC_ALL=C systemctl list-timers"))
```
see above, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        self.assertIn("Fri 2020-04-10 02:40", m.execute("LC_ALL=C systemctl list-timers"))
+        m.execute("timedatectl set-time '2020-04-10 03:00:00'")
+        b.wait_present(svc_sel('daily_timer.timer'))
+        self.assertIn("Fri 2020-04-10 21:15", m.execute("LC_ALL=C systemctl list-timers"))
```
see above, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        m.execute("systemctl start hourly_timer.timer")
+        b.wait_present(svc_sel('hourly_timer.timer'))
+        b.wait_in_text(svc_sel('hourly_timer.timer'), "Hourly timer")
+        self.assertIn("Fri 2020-04-10 03:05", m.execute("LC_ALL=C systemctl list-timers"))
```
see above, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        self.assertIn("Fri 2020-04-10 03:05", m.execute("LC_ALL=C systemctl list-timers"))
+        m.execute("timedatectl set-time '2020-04-10 03:07:00'")
+        b.wait_present(svc_sel('hourly_timer.timer'))
+        self.assertIn("Fri 2020-04-10 03:26", m.execute("LC_ALL=C systemctl list-timers"))
```
see above, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        m.execute("timedatectl set-time '2020-04-10 04:00:00'")
+        b.wait_present(svc_sel('hourly_timer.timer'))
+        # checks if timer runs on next hour at 5 min and 26 min
+        self.assertIn("Fri 2020-04-10 04:05", m.execute("LC_ALL=C systemctl list-timers"))
```
see above, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        self.assertIn("Fri 2020-04-10 04:05", m.execute("LC_ALL=C systemctl list-timers"))
+        m.execute("timedatectl set-time '2020-04-10 04:10:00'")
+        b.wait_present(svc_sel('hourly_timer.timer'))
+        self.assertIn("Fri 2020-04-10 04:26", m.execute("LC_ALL=C systemctl list-timers"))
```
see above, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        m.execute("timedatectl set-time '2020-04-10 04:10:00'")
+        m.execute("systemctl start no_repeat_timer.timer")
+        b.wait_present(svc_sel('no_repeat_timer.timer'))
+        self.assertIn("Fri 2020-04-10 23:59", m.execute("LC_ALL=C systemctl list-timers"))
```
see above, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,270 @@ WantedBy=default.target
+        m.execute("systemctl start yearly_timer.timer")
+        b.wait_present(svc_sel('yearly_timer.timer'))
+        b.wait_in_text(svc_sel('yearly_timer.timer'), "Yearly timer")
+        self.assertIn("Wed 2020-01-01 16:00", m.execute("LC_ALL=C systemctl list-timers"))
```
see below, probably a race condition (maybe it would be good to write a function to cover waiting for this info)

dperpeet commented:
There are some potential races in the tests - otherwise I think they look good.
You have to be careful when checking conditions - if you test them directly, then you know that they are in effect immediately after the previous statements. Updating timer files isn't guaranteed to happen immediately, so some waiting is probably called for.
It might be good to add a helper function that sets the time and waits for a specific timer to show up inside cockpit (not cli).

harishanand95 commented on test/verify/check-services:
```diff
@@ -194,6 +195,9 @@ WantedBy=default.target
             return 'tr[data-goto-unit="%s"]' % service
+        def wait_systemctl_timer(time):
+            with testvm.Timeout(seconds=10, error_message="Timeout while waiting for systemctl list-timers"):
+                m.execute("cmd='systemctl list-timers'; until $cmd | grep -m 1 '%s'; do sleep 1; done" % time)
```
m.execute("cmd='systemctl list-timers'; until $cmd | grep -m 1 '%s'; do _sleep 1_; done" % time) would loop and check if systemctl list-timers has the updated results. This keeps checking every second for a total of 10sec after which timeout occurs. 
@dperpeet which is better sleep 1 or sleep 0.5, here?

dperpeet replied:
> @dperpeet which is better sleep 1 or sleep 0.5, here?

1 is fine

harishanand95 commented on test/verify/check-services:
```diff
@@ -397,16 +407,17 @@ WantedBy=default.target
         self.assertIn("Fri 2020-04-10 04:05", m.execute("LC_ALL=C systemctl list-timers"))
         m.execute("timedatectl set-time '2020-04-10 04:10:00'")
-        b.wait_present(svc_sel('hourly_timer.timer'))
+        wait_systemctl_timer("Fri 2020-04-10 04:26")
```
_b.wait_present(svc_sel('hourly_timer.timer'))_ was removed because I thought they would solve the race issue here. Since they didn't, I'm removing all such checks we had.

Reason why _b.wait_present(svc_sel('hourly_timer.timer'))_ could solve race was based on the idea that cockpit timer page refreshes after changing time using timedatectl.

dperpeet commented:
debian tests fail - needs fix from #4857

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,281 @@ WantedBy=default.target
+            return 'tr[data-goto-unit="%s"]' % service
+        def wait_systemctl_timer(time):
+            with testvm.Timeout(seconds=20, error_message="Timeout while waiting for systemctl list-timers"):
+                m.execute("sleep 1")
```
why do you need to sleep here? the next loops does that

harishanand95 replied:
I thought if we called `systemctl list-timer` 's in next line after 1 sec, then it can have the updated values. Let me test it locally in centos and fedora-atomic without the 1 sec and see if it works.

harishanand95 replied:
TESTS PASSED on centos and fedora-atomic locally.

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,281 @@ WantedBy=default.target
+        def svc_sel(service):
+            return 'tr[data-goto-unit="%s"]' % service
+        def wait_systemctl_timer(time):
+            with testvm.Timeout(seconds=20, error_message="Timeout while waiting for systemctl list-timers"):
```
Timeout is patched now, you can add `machine=m` to make sure it actually works with the ssh command

dperpeet commented on test/verify/check-services:
```diff
@@ -188,6 +189,281 @@ WantedBy=default.target
+        m.spawn("sync && sync && sync && sleep 0.1 && reboot", "reboot")
+        m.wait_reboot()
+        m.start_cockpit()
+        with testvm.Timeout(seconds=15, error_message="Timeout while waiting for boot timer to run"):
```
here you can also add `machine=m`

harishanand95 commented on pkg/systemd/init.js:
```diff
@@ -823,4 +829,484 @@ define([
+            fail(function(error) {
+                console.log(error);
+            });
+        console.log("#Timer file#\n" + timer_file);
```
@dperpeet before we merge this, we should remove these _console.log()_ lines. They give the contents of .service and .timer files.

dperpeet replied:
good point, I will do this during merge

dperpeet commented:
nice job, this will be part of Cockpit 0.118

harishanand95 commented:
great! thanks everyone!
